roll:
  localizationMode: false # for testing 'GT' accuracy
  saveMapDirectory:  "/home/lzq/ROLL_ws/logs/fastlio_loc2/map_pcd"           
  saveKeyframeMapDirectory:  "/home/lzq/ROLL_ws/logs/fastlio_loc2/keyframes" 
  loadKeyframeMapDirectory:  "/home/lzq/ROLL_ws/logs/fastlio_loc2/keyframes" 

  saveKeyframeMap: true 
  savePose: true # for path recording purposes /reloc gt, save it to saveMapDirectory
  savePCD: true
  saveRawCloud: false

  saveLog: true
  relocSuccess: false

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius(index would overflow if too big)
  globalMapVisualizationPoseDensity: 1.0       # 10 by default, meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 0.5         # 1 by default, meters, global map visualization cloud density

    # Surrounding map
  surroundingkeyframeAddingDistThreshold: 5.0  # 1.0 by default,meters, regulate keyframe adding threshold
  surroundingKeyframeDensity: 2.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 50.0         # 20 is not okay; meters, within n meters scan-to-map optimization (when loop closure disabled)
  # Topics
  pointCloudTopic: cloud_registered_body              # Point cloud data
  gpsTopic: fix1                 # GPS odometry topic from navsat, see module_navsat.launch file
  gtTopic: ground_truth

  # GPS Settings
  useImuHeadingInitialization: false           # if using GPS data, set to "true"
  useGpsElevation: false                      # false by default, if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 1.0                      # 0.1 would be too frequent, 1.0 by default, m^2, threshold for using GPS data

  # Frames
  lidarFrame: "lidar_link"
  mapFrame: "odom"
  # NCLT HDL-32: 
  sensor: velodyne                            # lidar sensor type, either 'velodyne' or 'ouster'
  N_SCAN: 32                                 ## number of lidar channel (i.e., 16, 32, 64, 128)
  Horizon_SCAN: 1800                          ## lidar horizontal resolution (Velodyne:1800, Ouster:512,1024,2048,robosense: 2000)
  downsampleRate: 2                           #  default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1

  lidarMinRange: 0.2                          # default: 1.0, minimum lidar range to be used
  lidarMaxRange: 150.0                      ## default: 1000.0, maximum lidar range to be used

  # LOAM feature threshold
  edgeThreshold: 0.1 # changed the way getting smoothness
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # voxel filter paprams
  odometrySurfLeafSize: 0.4                     # default: 0.4 - outdoor, 0.2 - indoor
  mappingCornerLeafSize: 0.2                    # default: 0.2 - outdoor, 0.1 - indoor
  mappingSurfLeafSize: 0.4                      # default: 0.4 - outdoor, 0.2 - indoor

  # CPU Params
  numberOfCores: 4                              # number of cores for mapping optimization


  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                    # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 15.0             # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 30.0           # 30 by default,seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 25                  # number of hostory key frames will be fused into a submap for loop closure
  # 0.3 by default, but for indoor it can go wrong, actually mostly for good matches, fitnessScore is around 0.03 ;
  #icp threshold, the smaller the better alignment
  historyKeyframeFitnessScore: 0.2              


